File: .claude/sdk/custom-agent.ts
/**
 * Custom Agent using Claude Code SDK
 * This demonstrates how to create specialized agents with the SDK
 */

import { ClaudeCode, Agent, Tool } from '@anthropic/claude-code-sdk';

// Custom tool for project-specific operations
class ProjectAnalyzer implements Tool {
  name = 'project-analyzer';
  description = 'Analyzes project structure and dependencies';
  
  async execute(params: { path: string }): Promise<any> {
    // Custom analysis logic
    const analysis = {
      structure: await this.analyzeStructure(params.path),
      dependencies: await this.analyzeDependencies(params.path),
      complexity: await this.calculateComplexity(params.path)
    };
    return analysis;
  }
  
  private async analyzeStructure(path: string) {
    // Implementation
    return {};
  }
  
  private async analyzeDependencies(path: string) {
    // Implementation
    return {};
  }
  
  private async calculateComplexity(path: string) {
    // Implementation
    return { score: 0 };
  }
}

// Custom agent for architecture decisions
export class ArchitectureAgent extends Agent {
  name = 'architecture';
  description = 'Makes architecture decisions and refactoring suggestions';
  
  constructor() {
    super({
      tools: [
        new ProjectAnalyzer(),
        // Add more custom tools
      ],
      systemPrompt: `
        You are an expert software architect.
        Analyze codebases for:
        - Design patterns and anti-patterns
        - Performance bottlenecks
        - Scalability issues
        - Security vulnerabilities
        - Technical debt
        
        Provide actionable recommendations with:
        - Specific refactoring steps
        - Migration paths
        - Risk assessments
        - Implementation priorities
      `
    });
  }
  
  async analyzeArchitecture(projectPath: string) {
    const analysis = await this.execute(`
      Analyze the architecture of the project at ${projectPath}.
      Identify issues and provide recommendations.
    `);
    return analysis;
  }
  
  async suggestRefactoring(filePath: string) {
    const suggestions = await this.execute(`
      Analyze ${filePath} and suggest refactoring improvements.
      Focus on maintainability and performance.
    `);
    return suggestions;
  }
}

// Custom agent for data migration
export class MigrationAgent extends Agent {
  name = 'migration';
  description = 'Handles database migrations and data transformations';
  
  constructor() {
    super({
      systemPrompt: `
        You are a database migration expert.
        Handle:
        - Schema migrations
        - Data transformations
        - Rollback strategies
        - Zero-downtime deployments
        
        Always ensure:
        - Data integrity
        - Backward compatibility
        - Performance optimization
        - Proper indexing
      `
    });
  }
  
  async generateMigration(changes: string) {
    return await this.execute(`
      Generate a migration for: ${changes}
      Include both up and down migrations.
    `);
  }
  
  async validateMigration(migration: string) {
    return await this.execute(`
      Validate this migration for safety: ${migration}
      Check for data loss risks.
    `);
  }
}

// Custom agent for performance optimization
export class PerformanceAgent extends Agent {
  name = 'performance';
  description = 'Optimizes code and queries for performance';
  
  constructor() {
    super({
      systemPrompt: `
        You are a performance optimization specialist.
        Focus on:
        - Query optimization
        - Caching strategies
        - Algorithm complexity
        - Memory usage
        - Async operations
        
        Provide:
        - Benchmark comparisons
        - Optimization strategies
        - Trade-off analysis
        - Implementation code
      `
    });
  }
  
  async optimizeQuery(query: string) {
    return await this.execute(`
      Optimize this query: ${query}
      Provide the optimized version with explanation.
    `);
  }
  
  async analyzePerformance(code: string) {
    return await this.execute(`
      Analyze performance issues in: ${code}
      Suggest specific optimizations.
    `);
  }
}

// Agent orchestrator
export class AgentOrchestrator {
  private agents: Map<string, Agent>;
  
  constructor() {
    this.agents = new Map([
      ['architecture', new ArchitectureAgent()],
      ['migration', new MigrationAgent()],
      ['performance', new PerformanceAgent()],
    ]);
  }
  
  async executeTask(task: string) {
    // Determine which agent(s) to use
    const agent = this.selectAgent(task);
    
    // Execute with selected agent
    const result = await agent.execute(task);
    
    // Post-process if needed
    return this.processResult(result);
  }
  
  private selectAgent(task: string): Agent {
    // Logic to select appropriate agent based on task
    if (task.includes('migration') || task.includes('database')) {
      return this.agents.get('migration')!;
    }
    if (task.includes('performance') || task.includes('optimize')) {
      return this.agents.get('performance')!;
    }
    return this.agents.get('architecture')!;
  }
  
  private processResult(result: any) {
    // Post-processing logic
    return result;
  }
}

// Export for use in projects
export default {
  ArchitectureAgent,
  MigrationAgent,
  PerformanceAgent,
  AgentOrchestrator
};

File: .claude/sdk/custom-agent.py
"""
Custom Agent using Claude Code SDK (Python)
This demonstrates how to create specialized agents with the SDK
"""

from typing import Dict, Any, List, Optional
from anthropic import AsyncAnthropic
import asyncio
import json


class Tool:
    """Base class for custom tools"""
    
    def __init__(self, name: str, description: str):
        self.name = name
        self.description = description
    
    async def execute(self, params: Dict[str, Any]) -> Any:
        raise NotImplementedError


class ProjectAnalyzer(Tool):
    """Custom tool for project analysis"""
    
    def __init__(self):
        super().__init__(
            name="project-analyzer",
            description="Analyzes project structure and dependencies"
        )
    
    async def execute(self, params: Dict[str, Any]) -> Dict[str, Any]:
        path = params.get("path", ".")
        
        analysis = {
            "structure": await self.analyze_structure(path),
            "dependencies": await self.analyze_dependencies(path),
            "complexity": await self.calculate_complexity(path),
            "tech_debt": await self.assess_tech_debt(path)
        }
        
        return analysis
    
    async def analyze_structure(self, path: str) -> Dict:
        # Implement structure analysis
        return {"modules": 0, "layers": []}
    
    async def analyze_dependencies(self, path: str) -> Dict:
        # Implement dependency analysis
        return {"external": [], "internal": []}
    
    async def calculate_complexity(self, path: str) -> Dict:
        # Implement complexity calculation
        return {"cyclomatic": 0, "cognitive": 0}
    
    async def assess_tech_debt(self, path: str) -> Dict:
        # Implement tech debt assessment
        return {"score": 0, "items": []}


class Agent:
    """Base class for custom agents"""
    
    def __init__(
        self,
        name: str,
        description: str,
        system_prompt: str,
        tools: Optional[List[Tool]] = None
    ):
        self.name = name
        self.description = description
        self.system_prompt = system_prompt
        self.tools = tools or []
        self.client = AsyncAnthropic()
    
    async def execute(self, task: str) -> str:
        """Execute a task using Claude"""
        
        # Build tool descriptions for the prompt
        tool_descriptions = self._build_tool_descriptions()
        
        # Create the full prompt
        prompt = f"""
{self.system_prompt}

Available tools:
{tool_descriptions}

Task: {task}
"""
        
        # Get response from Claude
        response = await self.client.messages.create(
            model="claude-3-opus-20240229",
            max_tokens=4000,
            messages=[{"role": "user", "content": prompt}]
        )
        
        return response.content[0].text
    
    def _build_tool_descriptions(self) -> str:
        """Build descriptions of available tools"""
        if not self.tools:
            return "No special tools available."
        
        descriptions = []
        for tool in self.tools:
            descriptions.append(f"- {tool.name}: {tool.description}")
        
        return "\n".join(descriptions)


class ArchitectureAgent(Agent):
    """Agent for architecture decisions and refactoring"""
    
    def __init__(self):
        super().__init__(
            name="architecture",
            description="Makes architecture decisions and refactoring suggestions",
            system_prompt="""
You are an expert software architect.

Analyze codebases for:
- Design patterns and anti-patterns
- Performance bottlenecks
- Scalability issues
- Security vulnerabilities
- Technical debt

Provide actionable recommendations with:
- Specific refactoring steps
- Migration paths
- Risk assessments
- Implementation priorities
""",
            tools=[ProjectAnalyzer()]
        )
    
    async def analyze_architecture(self, project_path: str) -> Dict:
        """Analyze project architecture"""
        
        # Use the project analyzer tool
        analyzer = self.tools[0]
        analysis = await analyzer.execute({"path": project_path})
        
        # Get Claude's assessment
        assessment = await self.execute(f"""
Analyze this project structure and provide architecture recommendations:
{json.dumps(analysis, indent=2)}
""")
        
        return {
            "analysis": analysis,
            "recommendations": assessment
        }
    
    async def suggest_refactoring(self, file_path: str, code: str) -> str:
        """Suggest refactoring for specific code"""
        
        return await self.execute(f"""
Analyze this code and suggest refactoring improvements:

File: {file_path}
Code:
{code}

Focus on maintainability, performance, and best practices.
""")


class MigrationAgent(Agent):
    """Agent for database migrations and data transformations"""
    
    def __init__(self):
        super().__init__(
            name="migration",
            description="Handles database migrations and data transformations",
            system_prompt="""
You are a database migration expert.

Handle:
- Schema migrations
- Data transformations
- Rollback strategies
- Zero-downtime deployments

Always ensure:
- Data integrity
- Backward compatibility
- Performance optimization
- Proper indexing
"""
        )
    
    async def generate_migration(self, changes: str) -> Dict[str, str]:
        """Generate up and down migrations"""
        
        result = await self.execute(f"""
Generate SQL migrations for these changes:
{changes}

Provide:
1. Up migration (apply changes)
2. Down migration (rollback)
3. Safety checks
4. Performance considerations
""")
        
        # Parse result into structured format
        # This is simplified - real implementation would parse properly
        return {
            "up": result,
            "down": "-- Rollback migration",
            "checks": "-- Safety checks"
        }
    
    async def validate_migration(self, migration: str) -> Dict[str, Any]:
        """Validate migration for safety"""
        
        result = await self.execute(f"""
Validate this migration for safety:

{migration}

Check for:
- Data loss risks
- Locking issues
- Performance impact
- Rollback feasibility
""")
        
        return {
            "is_safe": True,  # Would parse from result
            "risks": [],
            "recommendations": result
        }


class PerformanceAgent(Agent):
    """Agent for performance optimization"""
    
    def __init__(self):
        super().__init__(
            name="performance",
            description="Optimizes code and queries for performance",
            system_prompt="""
You are a performance optimization specialist.

Focus on:
- Query optimization
- Caching strategies
- Algorithm complexity
- Memory usage
- Async operations

Provide:
- Benchmark comparisons
- Optimization strategies
- Trade-off analysis
- Implementation code
"""
        )
    
    async def optimize_query(self, query: str, schema: Optional[str] = None) -> Dict:
        """Optimize database query"""
        
        prompt = f"Optimize this query:\n{query}"
        if schema:
            prompt += f"\n\nSchema:\n{schema}"
        
        optimized = await self.execute(prompt)
        
        return {
            "original": query,
            "optimized": optimized,
            "explanation": "Query optimized for performance"
        }
    
    async def analyze_performance(self, code: str, language: str = "python") -> Dict:
        """Analyze code for performance issues"""
        
        result = await self.execute(f"""
Analyze this {language} code for performance issues:

{code}

Identify:
- Bottlenecks
- Inefficient algorithms
- Memory leaks
- Optimization opportunities
""")
        
        return {
            "issues": [],  # Would parse from result
            "suggestions": result,
            "priority": "medium"
        }


class SecurityAgent(Agent):
    """Agent for security analysis and fixes"""
    
    def __init__(self):
        super().__init__(
            name="security",
            description="Analyzes and fixes security vulnerabilities",
            system_prompt="""
You are a security expert.

Check for:
- OWASP Top 10 vulnerabilities
- Authentication/authorization issues
- Data exposure risks
- Injection attacks
- Cryptographic weaknesses

Provide:
- Vulnerability details
- Risk assessment
- Fix implementation
- Prevention strategies
"""
        )
    
    async def scan_vulnerabilities(self, code: str) -> List[Dict]:
        """Scan code for security vulnerabilities"""
        
        result = await self.execute(f"""
Scan this code for security vulnerabilities:

{code}

Report all findings with severity levels.
""")
        
        # Would parse structured vulnerabilities from result
        return [
            {
                "type": "example",
                "severity": "low",
                "description": result
            }
        ]
    
    async def fix_vulnerability(self, code: str, vulnerability: str) -> str:
        """Fix specific vulnerability in code"""
        
        return await self.execute(f"""
Fix this vulnerability in the code:

Vulnerability: {vulnerability}

Code:
{code}

Provide the secure version.
""")


class AgentOrchestrator:
    """Orchestrates multiple agents for complex tasks"""
    
    def __init__(self):
        self.agents = {
            "architecture": ArchitectureAgent(),
            "migration": MigrationAgent(),
            "performance": PerformanceAgent(),
            "security": SecurityAgent()
        }
    
    async def execute_task(self, task: str) -> Any:
        """Execute task with appropriate agent(s)"""
        
        # Determine which agent to use
        agent = self._select_agent(task)
        
        # Execute task
        result = await agent.execute(task)
        
        # Post-process if needed
        return self._process_result(result, agent.name)
    
    def _select_agent(self, task: str) -> Agent:
        """Select appropriate agent based on task"""
        
        task_lower = task.lower()
        
        if any(word in task_lower for word in ["migrate", "database", "schema"]):
            return self.agents["migration"]
        
        if any(word in task_lower for word in ["performance", "optimize", "slow"]):
            return self.agents["performance"]
        
        if any(word in task_lower for word in ["security", "vulnerability", "auth"]):
            return self.agents["security"]
        
        # Default to architecture agent
        return self.agents["architecture"]
    
    def _process_result(self, result: Any, agent_name: str) -> Dict:
        """Process and format agent result"""
        
        return {
            "agent": agent_name,
            "result": result,
            "timestamp": asyncio.get_event_loop().time()
        }
    
    async def execute_pipeline(self, tasks: List[str]) -> List[Dict]:
        """Execute multiple tasks in sequence"""
        
        results = []
        
        for task in tasks:
            result = await self.execute_task(task)
            results.append(result)
            
            # Use result for next task if needed
            # This enables chaining of agent actions
        
        return results


# Example usage
async def main():
    """Example of using custom agents"""
    
    # Single agent usage
    arch_agent = ArchitectureAgent()
    analysis = await arch_agent.analyze_architecture("./src")
    print(f"Architecture analysis: {analysis}")
    
    # Orchestrator usage
    orchestrator = AgentOrchestrator()
    
    # Complex task pipeline
    pipeline_tasks = [
        "Analyze the security of the authentication system",
        "Optimize the user query performance",
        "Generate migration to add audit logging"
    ]
    
    results = await orchestrator.execute_pipeline(pipeline_tasks)
    
    for i, result in enumerate(results):
        print(f"\nTask {i+1} - Agent: {result['agent']}")
        print(f"Result: {result['result'][:200]}...")  # Truncate for display


if __name__ == "__main__":
    # Run example
    asyncio.run(main())